docs/design/features/native-hosting.md

```diff
@@ -37,6 +37,69 @@ This document focuses on hosting which cooperates with the .NET SDK and consumes
+The end result of this step should be an initialized host context for a given managed code input which the native host can use to load the managed code and execute it. The step should not actually load any runtime or managed code into the process. Finally this step must handle both processes which don't have any runtime loaded as well as processes which already have runtime loaded.
+
+Functionality performed by this step:
+* Locating and understanding the managed code input (app/component)
+* Resolving frameworks or handling self-contained
+* Resolving dependencies (`.deps.json` processing)
+* Populating runtime properties
+
+The vision is to expose a set of APIs in the form of `hostfxr_initialize_for_...` which would all perform the above functionality and would differ on how the managed code is located and used. So there should be a different API for loading applications and components at the very least. Going forward we can add API which can initialize the context from in-memory configurations, or API which can initialize an "empty" context (no custom code involved, just frameworks) and so on.
+
+First step in this direction is the introduction of these APIs as described below:
+* `hostfxr_initialize_for_dotnet_command_line` - this is the initialize for an application. The API is a bit more complicated as it allows for command line parsing as well. Eventually we might want to add another "initialize app" API without the command line support.
+* `hostfxr_initialize_for_runtime_config` - this is the initialize for a component (naming is not ideal, but can't be changed anymore).
+
+After this step it should be possible for the native host to inspect and potentially modify runtime properties. APIs like `hostfxr_get_runtime_property_value` and similar described below are an example of this.
+
+### Loading managed code
+This step might be "virtual" in the sense that it's a way to setup the context for the desired functionality. In several scenarios it's not practical to be able to perform this step in separation from the "execute code" step below. The goal of this step is to determine which managed code to load and where it should be loaded, and to setup the correct inputs so that runtime can perform the right dependency resolution.
+
+The main functionality this step should provide is to determine which assembly load context will be used for loading which code:
+* Load all managed code into the default load context
+* Load the specified custom managed code into an isolated load context
+
+Eventually we could also add functionality for example to setup the isolated load context for unloadability and so on.
+
+Note that there are no APIs proposed in this document which would only perform this step for now. All the APIs so far fold this step into the next one and perform both loading and executing of the managed code in one step. We should try to make this step more explicit in the future to allow for greater flexibility.
+
```

Comment:
Do we want to call out that modifying the context (e.g. runtime properties) is no longer possible after this step? (Since we point out it is possible after the previous one)